File: solutions/csharp/forth/2/Context.cs
public class Context
{
    public Dictionary<string, string> InstructionsByWord { get; init; } = new(StringComparer.OrdinalIgnoreCase);

    private static readonly CultureInfo _invariantCulture = CultureInfo.InvariantCulture;

    [SuppressMessage("Design", "CA1002:Do not expose generic lists", Justification = "<Pending>")]
    public List<int> Stack { get; init; } = new();

    public override string ToString()
        => string.Join(' ', Stack);

    public Context DefineHandler(Group group)
    {
        _ = group ?? throw new ArgumentNullException(nameof(group));
        List<string> definition = group.Value.Split(' ')
            .Where(x => !string.IsNullOrEmpty(x))
            .ToList()
            .RequireAtLeastTwoValues();

        string word = definition.Shift();
        if (int.TryParse(word, out int _))
        {
            throw new InvalidOperationException("Redefining numbers not allowed.");
        }

        InstructionsByWord[word] = CreateInstruction(definition);
        return this;
    }

    private string CreateInstruction(List<string> definition)
    {
        StringBuilder builder = new();
        foreach (string word in definition)
        {
            _ = InstructionsByWord.TryGetValue(word, out string instruction)
                ? builder.Append(_invariantCulture, $"{instruction} ")
                : builder.Append(_invariantCulture, $"{word} ");
        }

        return builder.ToString()
            .Trim();
    }

    public Context WithStack(Action<List<int>> function)
    {
        _ = function ?? throw new ArgumentNullException(nameof(function));
        function(Stack);
        return this;
    }
}

File: solutions/java/forth/1/src/main/java/Context.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.function.Consumer;
import java.util.stream.Stream;

public class Context {
    private final Map<String, String> instructionsByWord = new HashMap<>();
    public Map<String, String> getInstructionsByWord() {
        return instructionsByWord;
    }

    private final List<Integer> stack = new ArrayList<>();
    public List<Integer> getStack() {
        return stack;
    }

    @Override
    public String toString() {
        return String.join(
                " ",
                stack.stream()
                     .map(String::valueOf)
                     .toList()
        );
    }

    public Context defineHandler(String definition) {
        if (definition == null) {
            throw new NullPointerException("group");
        }

        List<String> words = Stream.of(definition.split(" "))
                                   .filter(word -> !word.isEmpty())
                                   .collect(java.util.stream.Collectors.toList());
        ListUtilities.requireAtLeastTwoValues("Defining", words);
        String word = ListUtilities.shift(words);
        if (isNumber(word)) {
            throw new IllegalArgumentException("Cannot redefine numbers");
        }

        instructionsByWord.put(
                word.toLowerCase(Locale.ROOT),
                createInstruction(words)
        );
        return this;
    }

    private String createInstruction(List<String> definition) {
        StringBuilder builder = new StringBuilder();
        definition.forEach(word -> {
            String instruction = instructionsByWord.get(
                    word.toLowerCase(Locale.ROOT)
            );

            builder.append(instruction != null ? instruction : word)
                   .append(" ");
        });

        return builder.toString()
                      .trim();
    }

    public Context withStack(Consumer<List<Integer>> function) {
        if (function == null) {
            throw new NullPointerException("function");
        }

        function.accept(stack);
        return this;
    }

    private boolean isNumber(String value) {
        try {
            Integer.parseInt(value);
            return true;
        }
        catch (NumberFormatException exception) {
            return false;
        }
    }
}